tests/core/test_datamodules.py

```diff
@@ -399,14 +426,23 @@ def transfer_batch_to_device(self, data, device):
     trainer = Trainer(gpus=1)
     # running .fit() would require us to implement custom data loaders, we mock the model reference instead
     trainer.get_model = MagicMock(return_value=model)
+    if is_overridden('on_before_batch_transfer', dm):
+        model.on_before_batch_transfer = dm.on_before_batch_transfer
     if is_overridden('transfer_batch_to_device', dm):
         model.transfer_batch_to_device = dm.transfer_batch_to_device
+    if is_overridden('on_after_batch_transfer', dm):
+        model.on_after_batch_transfer = dm.on_after_batch_transfer
 
     trainer.accelerator_backend = GPUAccelerator(trainer)
     batch_gpu = trainer.accelerator_backend.batch_to_device(batch, torch.device('cuda:0'))
-    expected = torch.device('cuda', 0)
-    assert dm.hook_called
-    assert batch_gpu.samples.device == batch_gpu.targets.device == expected
+    expected_device = torch.device('cuda', 0)
+
+    assert dm.on_before_batch_transfer_hook_rank == 0
+    assert dm.transfer_batch_to_device_hook_rank == 1
+    assert dm.on_after_batch_transfer_hook_rank == 2
+    assert batch_gpu.samples.device == batch_gpu.targets.device == expected_device
+    assert torch.allclose(batch_gpu.samples.cpu(), torch.ones(5, 28))
+    assert torch.allclose(batch_gpu.targets.cpu(), torch.ones(5, 1, dtype=torch.long) * 2)
```

Comment:
we can simplify this:

```suggestion
    assert all(batch_gpu.samples.cpu() == 1)
    assert all(batch_gpu.targets.cpu() == 2)
```

assuming you don't need to test the shape or dtype

Reply:
let's keep it as it is just to be super-sure